server/matchmaker/search.py

```diff
@@ -234,4 +235,79 @@ def __repr__(self):
         return f"Search({[p.login for p in self.players]})"
 
 
+class CombinedSearch(Search):
+    def __init__(self, *searches: Search):
+        assert searches
+        rating_type = searches[0].rating_type
+        assert all(map(lambda s: s.rating_type == rating_type, searches))
+
+        self.rating_type = rating_type
+        self.searches = searches
+
+    @property
+    def players(self) -> List[Player]:
+        return list(itertools.chain(*[s.players for s in self.searches]))
+
+    @property
+    def ratings(self):
+        return list(itertools.chain(*[s.ratings for s in self.searches]))
+
+    @property
+    def raw_ratings(self):
+        return list(itertools.chain(*[s.raw_ratings for s in self.searches]))
+
+    @property
+    def failed_matching_attempts(self) -> List[int]:
+        """Used for logging so returning a different type here is fine"""
+        return [search.failed_matching_attempts for search in self.searches]
+
+    def register_failed_matching_attempt(self):
+        for search in self.searches:
+            search.register_failed_matching_attempt()
+
+    @property
+    def match_threshold(self) -> float:
+        """
+        Defines the threshold for game quality
+        """
+        return min(s.match_threshold for s in self.searches)
+
+    @property
+    def is_matched(self) -> bool:
+        return all(s.is_matched for s in self.searches)
+
+    def done(self) -> bool:
+        return all(s.done() for s in self.searches)
+
+    @property
```

Comment:
They are making `CombinedSearch` adhere to the `Search` interface, where you call `search.is_matched`, not `search.is_matched()` or `search.get_whether_matched()` etc, aren't they?

Reply:
Removing 2 brackets is not the "win" situation. It's just "let's hide some code after that property, and when it'll be fucked up, you'll search for problem, and you'll not know, that access too some variable will call this method, you'll just continue with reading without jumping here"... GL HF :see_no_evil: 

Properties sohuld be used when you are rewritting old interface, to new one, and want to have code backward compatibility. And nothing else like "i hide private variables, etc..", we're in python. :-)

Reply:
Yea it needs to have the same API of `Search`, and of course `CombinedSearch` needs to do more stuff for each check because it needs to make a "super" call to all contained searches.